File: engine/tests/smoke.test.js
/**
 * AIde Engine - Smoke Test
 *
 * Verifies that the JS engine build is functional.
 * Tests the full cycle: empty state → events → reduce → render → parse
 */

const { test } = require('node:test');
const assert = require('node:assert');
const { emptyState, reduce, replay, render, parseAideHtml } = require('../builds/engine.js');

// Helper to create events
function makeEvent(seq, type, payload) {
  return {
    id: `evt_${seq}`,
    sequence: seq,
    timestamp: new Date().toISOString(),
    actor: 'test',
    source: 'test',
    type,
    payload,
  };
}

test('emptyState returns valid initial state', () => {
  const state = emptyState();

  assert.strictEqual(state.version, 1);
  assert.deepStrictEqual(state.collections, {});
  assert.deepStrictEqual(state.relationships, []);
  assert.ok('block_root' in state.blocks);
  assert.strictEqual(state.blocks.block_root.type, 'root');
});

test('reduce applies collection.create', () => {
  const state = emptyState();
  const event = makeEvent(1, 'collection.create', {
    id: 'tasks',
    name: 'Tasks',
    schema: {
      title: 'string',
      done: 'bool',
      due: 'date?',
    },
  });

  const result = reduce(state, event);

  assert.strictEqual(result.applied, true);
  assert.strictEqual(result.error, null);
  assert.ok('tasks' in result.snapshot.collections);
  assert.strictEqual(result.snapshot.collections.tasks.name, 'Tasks');
  assert.strictEqual(result.snapshot.collections.tasks.schema.title, 'string');
});

test('reduce applies entity.create', () => {
  let state = emptyState();

  // Create collection first
  state = reduce(state, makeEvent(1, 'collection.create', {
    id: 'tasks',
    name: 'Tasks',
    schema: { title: 'string', done: 'bool' },
  })).snapshot;

  // Create entity
  const result = reduce(state, makeEvent(2, 'entity.create', {
    collection: 'tasks',
    id: 'task_1',
    fields: { title: 'Buy milk', done: false },
  }));

  assert.strictEqual(result.applied, true);
  const entity = result.snapshot.collections.tasks.entities.task_1;
  assert.strictEqual(entity.title, 'Buy milk');
  assert.strictEqual(entity.done, false);
  assert.strictEqual(entity._removed, false);
});

test('reduce rejects entity.create for missing collection', () => {
  const state = emptyState();
  const result = reduce(state, makeEvent(1, 'entity.create', {
    collection: 'nonexistent',
    id: 'item_1',
    fields: { name: 'Test' },
  }));

  assert.strictEqual(result.applied, false);
  assert.ok(result.error.includes('COLLECTION_NOT_FOUND'));
});

test('replay builds state from event sequence', () => {
  const events = [
    makeEvent(1, 'collection.create', {
      id: 'players',
      name: 'Players',
      schema: { name: 'string', score: 'int' },
    }),
    makeEvent(2, 'entity.create', {
      collection: 'players',
      id: 'player_1',
      fields: { name: 'Alice', score: 100 },
    }),
    makeEvent(3, 'entity.create', {
      collection: 'players',
      id: 'player_2',
      fields: { name: 'Bob', score: 85 },
    }),
    makeEvent(4, 'entity.update', {
      ref: 'players/player_1',
      fields: { score: 150 },
    }),
  ];

  const state = replay(events);

  assert.ok('players' in state.collections);
  const players = state.collections.players.entities;
  assert.strictEqual(Object.keys(players).length, 2);
  assert.strictEqual(players.player_1.score, 150);
  assert.strictEqual(players.player_2.score, 85);
});

test('render produces valid HTML', () => {
  let state = emptyState();

  // Build up state
  state = reduce(state, makeEvent(1, 'meta.update', {
    title: 'Test Page',
  })).snapshot;

  state = reduce(state, makeEvent(2, 'block.set', {
    id: 'block_heading',
    type: 'heading',
    parent: 'block_root',
    props: { level: 1, content: 'Welcome' },
  })).snapshot;

  const blueprint = { identity: 'test', voice: 'friendly', prompt: 'test' };
  const html = render(state, blueprint, []);

  assert.ok(html.includes('<!DOCTYPE html>'));
  assert.ok(html.includes('<title>Test Page</title>'));
  assert.ok(html.includes('Welcome'));
  assert.ok(html.includes('application/aide+json'));
  assert.ok(html.includes('application/aide-blueprint+json'));
});

test('parseAideHtml extracts embedded data', () => {
  const state = emptyState();
  const blueprint = { identity: 'test', voice: 'friendly', prompt: 'test' };
  const events = [
    makeEvent(1, 'meta.update', { title: 'Round Trip' }),
  ];

  const html = render(state, blueprint, events);
  const parsed = parseAideHtml(html);

  assert.strictEqual(parsed.blueprint.identity, 'test');
  assert.strictEqual(parsed.snapshot.version, 1);
  assert.ok(Array.isArray(parsed.events));
});

test('full round-trip: events → state → html → parse', () => {
  // Build complex state
  const events = [
    makeEvent(1, 'meta.update', { title: 'Poker League' }),
    makeEvent(2, 'collection.create', {
      id: 'roster',
      name: 'Players',
      schema: { name: 'string', status: 'string' },
    }),
    makeEvent(3, 'entity.create', {
      collection: 'roster',
      id: 'player_mike',
      fields: { name: 'Mike', status: 'active' },
    }),
    makeEvent(4, 'entity.create', {
      collection: 'roster',
      id: 'player_dave',
      fields: { name: 'Dave', status: 'active' },
    }),
    makeEvent(5, 'view.create', {
      id: 'roster_view',
      type: 'table',
      source: 'roster',
      config: { show_fields: ['name', 'status'] },
    }),
    makeEvent(6, 'block.set', {
      id: 'block_title',
      type: 'heading',
      parent: 'block_root',
      props: { level: 1, content: 'Poker League' },
    }),
    makeEvent(7, 'block.set', {
      id: 'block_roster',
      type: 'collection_view',
      parent: 'block_root',
      props: { view_id: 'roster_view' },
    }),
  ];

  const state = replay(events);
  const blueprint = { identity: 'poker', voice: 'casual', prompt: '' };
  const html = render(state, blueprint, events);

  // Parse it back
  const parsed = parseAideHtml(html);

  // Verify integrity
  assert.strictEqual(parsed.snapshot.meta.title, 'Poker League');
  assert.ok('roster' in parsed.snapshot.collections);
  assert.strictEqual(Object.keys(parsed.snapshot.collections.roster.entities).length, 2);
  assert.ok('roster_view' in parsed.snapshot.views);
  assert.strictEqual(parsed.events.length, events.length);
});

console.log('Smoke test complete.');

File: engine/kernel/smoke_test.py
"""
End-to-end smoke test for AIde kernel.

Tests the full flow: events → reducer → renderer → HTML
"""

import asyncio

from engine.kernel.assembly import AideAssembly, MemoryStorage, parse_aide_html
from engine.kernel.reducer import empty_state, reduce
from engine.kernel.renderer import render
from engine.kernel.types import Blueprint, Event, RenderOptions


async def main():
    print("🧪 AIde Kernel End-to-End Smoke Test")
    print("=" * 60)

    # Step 1: Create events
    print("\n1️⃣  Creating primitive events...")
    events = [
        Event(
            id="evt_1",
            sequence=1,
            timestamp="2024-01-01T00:00:00Z",
            actor="test",
            source="smoke_test",
            type="meta.update",
            payload={"title": "Grocery List"},
        ),
        Event(
            id="evt_2",
            sequence=2,
            timestamp="2024-01-01T00:00:01Z",
            actor="test",
            source="smoke_test",
            type="collection.create",
            payload={
                "id": "groceries",
                "name": "Items",
                "schema": {
                    "name": "string",
                    "checked": "bool",
                },
                "primary_field": "name",
            },
        ),
        Event(
            id="evt_3",
            sequence=3,
            timestamp="2024-01-01T00:00:02Z",
            actor="test",
            source="smoke_test",
            type="entity.create",
            payload={
                "collection": "groceries",
                "id": "item_milk",
                "fields": {"name": "Milk", "checked": False},
            },
        ),
        Event(
            id="evt_4",
            sequence=4,
            timestamp="2024-01-01T00:00:03Z",
            actor="test",
            source="smoke_test",
            type="entity.create",
            payload={
                "collection": "groceries",
                "id": "item_eggs",
                "fields": {"name": "Eggs", "checked": False},
            },
        ),
        Event(
            id="evt_5",
            sequence=5,
            timestamp="2024-01-01T00:00:04Z",
            actor="test",
            source="smoke_test",
            type="view.create",
            payload={
                "id": "view_main",
                "source": "groceries",
                "type": "list",
            },
        ),
        Event(
            id="evt_6",
            sequence=6,
            timestamp="2024-01-01T00:00:05Z",
            actor="test",
            source="smoke_test",
            type="block.set",
            payload={
                "id": "block_header",
                "type": "heading",
                "content": "My Groceries",
            },
        ),
        Event(
            id="evt_7",
            sequence=7,
            timestamp="2024-01-01T00:00:06Z",
            actor="test",
            source="smoke_test",
            type="block.set",
            payload={
                "id": "block_list",
                "type": "collection_view",
                "view_id": "view_main",
            },
        ),
        Event(
            id="evt_8",
            sequence=8,
            timestamp="2024-01-01T00:00:07Z",
            actor="test",
            source="smoke_test",
            type="block.set",
            payload={
                "id": "block_root",
                "type": "root",
                "children": ["block_header"],
            },
        ),
        Event(
            id="evt_9",
            sequence=9,
            timestamp="2024-01-01T00:00:08Z",
            actor="test",
            source="smoke_test",
            type="block.set",
            payload={
                "id": "block_root",
                "type": "root",
                "children": ["block_header", "block_list"],
            },
        ),
    ]
    print(f"   ✓ Created {len(events)} events")

    # Step 2: Apply events through reducer
    print("\n2️⃣  Applying events through reducer...")
    snapshot = empty_state()
    applied_count = 0
    for event in events:
        result = reduce(snapshot, event)
        if result.applied:
            snapshot = result.snapshot
            applied_count += 1
        else:
            print(f"   ✗ Event {event.id} rejected: {result.error}")
    print(f"   ✓ Applied {applied_count}/{len(events)} events successfully")

    # Verify snapshot state
    assert "groceries" in snapshot["collections"], "Collection not created"
    assert "item_milk" in snapshot["collections"]["groceries"]["entities"], "Entity not created"
    assert snapshot["meta"]["title"] == "Grocery List", "Meta not set"
    print("   ✓ Snapshot state verified")

    # Step 3: Render to HTML
    print("\n3️⃣  Rendering snapshot to HTML...")
    blueprint = Blueprint(
        identity="Simple grocery list tracker",
        voice="State reflections only. No encouragement.",
        prompt="You are managing a grocery list.",
    )
    html = render(snapshot, blueprint, events, RenderOptions())
    print(f"   ✓ Generated HTML ({len(html)} bytes)")

    # Verify HTML structure
    assert "<!DOCTYPE html>" in html, "Missing DOCTYPE"
    assert '<html lang="en">' in html, "Missing HTML tag"
    assert "Grocery List" in html, "Missing title"
    assert "My Groceries" in html, "Missing heading"
    assert "Milk" in html, "Missing entity content"
    assert "Eggs" in html, "Missing entity content"
    print("   ✓ HTML structure verified")

    # Step 4: Parse HTML back
    print("\n4️⃣  Parsing HTML back to extract data...")
    parsed = parse_aide_html(html)
    assert parsed.snapshot is not None, "Snapshot not extracted"
    assert parsed.blueprint is not None, "Blueprint not extracted"
    assert len(parsed.events) == len(events), f"Events count mismatch: {len(parsed.events)} != {len(events)}"
    print("   ✓ Successfully parsed HTML")
    print(f"   ✓ Extracted {len(parsed.events)} events")

    # Step 5: Assembly layer round-trip
    print("\n5️⃣  Testing assembly layer round-trip...")
    storage = MemoryStorage()
    assembly = AideAssembly(storage)

    # Create new aide
    blueprint_new = Blueprint(
        identity="Grocery list for the week",
        voice="No first person. State reflections only.",
    )
    aide_file = await assembly.create(blueprint_new)
    aide_id = aide_file.aide_id
    print(f"   ✓ Created aide: {aide_id}")

    # Apply events
    result = await assembly.apply(aide_file, events)
    print(f"   ✓ Applied {len(result.applied)} events (rejected: {len(result.rejected)})")

    # Save
    await assembly.save(aide_file)
    print("   ✓ Saved to storage")

    # Load back
    loaded = await assembly.load(aide_id)
    assert loaded.aide_id == aide_id, "Aide ID mismatch"
    assert loaded.snapshot["meta"]["title"] == "Grocery List", "State not preserved"
    print("   ✓ Loaded from storage and verified")

    # Publish (pro tier - no footer)
    publish_url = await assembly.publish(loaded, slug="my-groceries", is_free_tier=False)
    assert publish_url is not None, "Publish failed"
    print(f"   ✓ Published to: {publish_url}")

    # Verify published HTML
    published_html = storage.published.get("my-groceries")
    assert published_html is not None, "Published HTML not found"
    assert "Made with AIde" not in published_html, "Pro tier should not have footer"
    print("   ✓ Published HTML verified (no footer for pro tier)")

    # Test free tier with footer
    publish_url_free = await assembly.publish(loaded, slug="my-groceries-free", is_free_tier=True)
    published_html_free = storage.published.get("my-groceries-free")
    assert "Made with AIde" in published_html_free, "Free tier should have footer"
    print("   ✓ Free tier footer verified")

    print("\n" + "=" * 60)
    print("✅ All smoke tests passed!")
    print("\nKernel is functioning correctly:")
    print("  • Events → Reducer → Snapshot ✓")
    print("  • Snapshot → Renderer → HTML ✓")
    print("  • HTML → Parser → Data ✓")
    print("  • Assembly layer (create/apply/save/load/publish) ✓")
    print("  • Round-trip integrity ✓")


if __name__ == "__main__":
    asyncio.run(main())
